msal/tests/Test.MSAL.NET.Unit/RequestsTests/InteractiveRequestTests.cs

```diff
@@ -235,6 +235,98 @@ public void RedirectUriContainsFragmentErrorTest()
             }
         }
 
+        [TestMethod]
+        [TestCategory("InteractiveRequestTests")]
+        public void OAuthClient_FailsWithServiceExceptionWhenItCannotParseJsonResponse()
+        {
+            Authority authority = Authority.CreateAuthority(TestConstants.AuthorityHomeTenant, false);
+
+            HttpMessageHandlerFactory.AddMockHandler(new MockHttpMessageHandler
+            {
+                Method = HttpMethod.Get,
+                ResponseMessage = MockHelpers.CreateTooManyRequestsNonJsonResponse() // returns a non json response
+            });
+
+            AuthenticationRequestParameters parameters = new AuthenticationRequestParameters()
+            {
+                Authority = authority,
+                ClientId = TestConstants.ClientId,
+                Scope = TestConstants.Scope,
+                TokenCache = null,
+                RequestContext = new RequestContext(new MsalLogger(Guid.NewGuid(), null)),
+            };
+            parameters.RedirectUri = new Uri("some://uri");
+
+            MockWebUI ui = new MockWebUI();
+
+            InteractiveRequest request = new InteractiveRequest(parameters,
+                TestConstants.ScopeForAnotherResource.ToArray(),
+                 TestConstants.DisplayableId,
+                UIBehavior.SelectAccount, ui);
+         
+
+            try
+            {
+                request.PreTokenRequestAsync(CancellationToken.None).Wait();
+                
+                Assert.Fail("MsalException should have been thrown here");
+            }
+            catch (Exception exc)
+            {
+                MsalServiceException serverEx = exc.InnerException as MsalServiceException;
+                Assert.IsNotNull(serverEx);
+                Assert.AreEqual(429, serverEx.StatusCode);
+                Assert.AreEqual(MockHelpers.TooManyRequestsContent, serverEx.ResponseBody);
+                Assert.AreEqual(MockHelpers.TestRetryAfterDuration, serverEx.Headers.RetryAfter.Delta);
+                Assert.AreEqual(CoreErrorCodes.OAuthNonJsonError, serverEx.ErrorCode);
+            }
+        }

```

Comment:
check that all the mocks were consumed?

Reply:
comment for future:  should we make the mock http handler stuff a disposable and auto-fail a test if all mocks aren't consumed?  might help us avoid having to add that every time..

Reply:
yes...definitely.

---
In reply to: [222787298](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1294#discussion_r222787298) [](ancestors = 222787298)